docs: Update the quick start guide (#930)
open-telemetry/opentelemetry-ruby: website_docs/quick_start.md

OpenTelemetry initialization:

```ruby
# config/initializers/opentelemetry.rb
require 'opentelemetry/sdk'
require 'opentelemetry/exporter/otlp'
require 'opentelemetry/instrumentation/all'
OpenTelemetry::SDK.configure do |c|
  c.service_name = '<YOUR_SERVICE_NAME>'
  c.use_all() # enables all instrumentation!
end
```

Now that you have setup your application to perform tracing you'll need to configure the SDK to export the traces to somewhere that will vizualize them for you.

### Exporting Traces

The following section assumes you are new to OpenTelemetry or do not currently use a vendor that supports distributed tracing using OTLP. Please refer to your vendor's product documentation if you would like to export your traces to a vendor for analysis and vizualization.

For the purposes of this tutorial you will configure an OpenTelemetry collector that will receive the traces and vizualize them using Jaeger or Zipkin UI.

First, start up an example system:

```bash
$> git clone git@github.com:open-telemetry/opentelemetry-ruby.git; \
    cd open-telemetry-ruby/examples/otel-collector; \

>        docker-compose up -d

Reviewer: ![MAGIC](https://user-images.githubusercontent.com/517302/132950565-07f4b182-c4d4-4909-a97a-45362025e3af.gif)

I'll test this out this weekend.

Author: You will have to switch to this branch in order for that to work.

---
title: "Ruby Quick Start Guide"
weight: 24
description: >
  <img width="35" src="https://raw.github.com/open-telemetry/opentelemetry.io/main/iconography/32x32/Ruby_SDK.svg"></img>
  A language-specific implementation of OpenTelemetry in Ruby.
---

## Quick Start

[OpenTelemetry for Ruby][repository] can be used to add automatic and manual instrumentation to your applications.
Automatic instrumentation is enabled by adding [instrumentation packages][auto-instrumentation].
Manual instrumentation can be added using the [OpenTelemetry API][manual-instrumentation].

### Requirements

These instructions will explain how to set up automatic and manual instrumentation for a Ruby service.
In order to follow along, you will need:

- Ruby 2.5 or higher
- Docker Compose

### Installation

> The first step add these gems to your Gemfile:

Reviewer 2: ```suggestion
The first step is to add these gems to your Gemfile:
```

```ruby
gem 'opentelemetry-sdk'
gem 'opentelemetry-exporter-otlp'
gem 'opentelemetry-instrumentation-all'
```

The inclusion of `opentelemetry-instrumentation-all` in the above list provides instrumentations for several frameworks such as Rails and Sinatra as well as database drivers and HTTP [libraries][auto-instrumentation].

### Initialization

The OpenTelemetry initialization needs to happen early in your application lifecycle.
For Rails applications, the usual way to initialize OpenTelemetry is in a Rails initializer.
For other Ruby services, perform this initialization as early as possible in the start-up process.

OpenTelemetry initialization:

```ruby
# config/initializers/opentelemetry.rb
require 'opentelemetry/sdk'
require 'opentelemetry/exporter/otlp'
require 'opentelemetry/instrumentation/all'
OpenTelemetry::SDK.configure do |c|
  c.service_name = '<YOUR_SERVICE_NAME>'
  c.use_all() # enables all instrumentation!
end
```

> Now that you have setup your application to perform tracing you'll need to configure the SDK to export the traces to somewhere that will vizualize them for you.

Reviewer 3: So ... technically one must configure the SDK to export the traces as part of initialization, not after (unless I've been doing it wrong? a distinct possibility, honestly).

Maybe we could say this:
```suggestion
Now that you have setup your application to perform tracing, you'll need to configure the SDK to export the traces somewhere. Our example loaded the `OTLP` exporter, which the SDK tries to use by default. Next, we'll use the OpenTelemetry Collector to receive these traces and the Jaeger UI to visualize them!
```

That might be too dense, though?

Author: How about this?

```suggestion
Now that you have setup your application to perform tracing, you'll need to configure the SDK to export the traces somewhere. Our example loaded the `OTLP` exporter, which the SDK tries to use by default. Next, we'll use the OpenTelemetry Collector to receive these traces and visualize them using Jaeger and Zipkin!
```

Reviewer 3: 👍 👍 👍 👍

For the purposes of this tutorial you will configure an OpenTelemetry collector that will receive the traces and vizualize them using Jaeger or Zipkin UI.

First, start up an example system:

```bash
$> git clone git@github.com:open-telemetry/opentelemetry-ruby.git; \
    cd open-telemetry-ruby/examples/otel-collector; \
       docker-compose up -d
```

Next, you'll have to let the SDK know where the collector endpoint is to receive traces.
Set the [`OTEL_EXPORTER_OTLP_ENDPOINT`][sdk-env] environment variable to `http://localhost:4318/v1/traces`:

```bash
export OTEL_EXPORTER_OTLP_ENDPOINT=http://localhost:4318/v1/traces
```

Now, start up your application and perform a few operations to generate tracing data, e.g. navigate around your web app or kick off background tasks.

Lastly, open a browser and navigate to the [Jaeger UI](http://localhost:16686) or [Zipkin UI](http://localhost:9411) and search for traces related to your service, which were generated by the auto-instrumentation features of OpenTelemetry!

### Achievement Unlocked: Tracing Enabled

> Adding tracing to a single service is a great first step and although auto-instrumenation provides quite a bit of insight on it's own, OpenTelemetry provides a few more features that will allow you gain even deeper insights!

Reviewer 2: ```suggestion
Adding tracing to a single service is a great first step and although auto-instrumenation provides quite a bit of insight on its own, OpenTelemetry provides a few more features that will allow you gain even deeper insights!
```

[Context Propagation][context-propagation] is perhaps one of the most powerful concepts in OpenTelemetry because it will upgrade your single service trace into a _distributed trace_, which makes it possible for OpenTelemetry vendors to visualize a request from end-to-end accross process and network boundaries.

> [Span Events][events] allow you to add a human-readable message on a span that represents "something happening" during it's lifetime.

Reviewer 2: ```suggestion
[Span Events][events] allow you to add a human-readable message on a span that represents "something happening" during its lifetime.
```
